jh/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+//! Implementation of JH cryptographic hash algorithms.
+//! The JH hash function was one of the submissions to SHA-3,
+//! the cryptographic hash algorithm competition organized by the NIST.
+//!
+//! There are 4 standard versions of the JH hash function:
+//!
+//! * `JH-224`
+//! * `JH-256`
+//! * `JH-384`
+//! * `JH-512`
+//!
+//! # Examples
+//!
+//! Output size of JH-256 is fixed, so its functionality is usually
+//! accessed via the `Digest` trait:
+//!
+//! ```
+//! use hex_literal::hex;
+//! use jh::{Digest, Jh256, digest::generic_array::typenum::U32};
+//!
+//! // create a JH-256 object
+//! let mut hasher = Jh256::new();
+//!
+//! // write input message
+//! hasher.update(b"hello");
+//!
+//! // read hash digest
+//! let result = hasher.finalize();
+//!
+//! assert_eq!(result[..], hex!("
+//!     94fd3f4c564957c6754265676bf8b244c707d3ffb294e18af1f2e4f9b8306089
+//! ")[..]);
+//! ```
+//! Also see [RustCrypto/hashes][2] readme.
+//!
+//! [1]: https://www3.ntu.edu.sg/home/wuhj/research/jh
+//! [2]: https://github.com/RustCrypto/hashes
+#![no_std]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/6ee8e381/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/6ee8e381/logo.svg"
+)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[doc(hidden)]
+pub mod compressor;
+mod consts;
+
+pub use digest::{self, Digest};
+
+use crate::compressor::Compressor;
+use core::fmt;
+use digest::{
+    block_buffer::Eager,
+    core_api::{
+        AlgorithmName, Block, Buffer, BufferKindUser, CoreWrapper, FixedOutputCore, UpdateCore,
+    },
+    crypto_common::{BlockSizeUser, OutputSizeUser},
+    generic_array::typenum::{Unsigned, U28, U32, U48, U64},
+    HashMarker, Output,
+};
+
+macro_rules! define_hasher {
```

Comment:
Instead of this macro you can implement `VariableOutputCore` for `JhCore` and then define the variants as:
```rust
pub type Jh256 = CoreWrapper<CtVariableCoreWrapper<JhCore, U32>>;
```
For reference you can take a look at the [`sha2`](https://github.com/RustCrypto/hashes/blob/master/sha2/src/core_api.rs) crate.